### PR TITLE
Rule S4225: Extension method should not extend on object - for VB.NET

### LINKED ISSUE #5823: Rule S4225: Extension method should not extend on object - for VB.NET
S4225

Creating an extension method that extends object is not recommended because it makes the method available on every type. Extensions should be applied at the most specialized level possible, and that is very unlikely to be object.

### PR DESCRIPTION
**Pull request by Corniel**
Fixes #5823

Originally implemented via #763. [Documentation update](https://github.com/SonarSource/rspec/pull/1063) pending.

### CODE REVIEW

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource any ETA?

**Comment by pavel-mikula-sonarsource** [comment]
My calendar is too busy last two weeks, it should be better next week.

**Comment by Corniel** [comment]
> My calendar is too busy last two weeks, it should be better next week.

I tried hard to reduce the effort needed by others. :)

**Comment by pavel-mikula-sonarsource** [comment]
Hi, I'm sorry, I won't be able to do this this week. And I'll be off the next one.

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExtensionMethodShouldNotExtendObject.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{
+    protected override ILanguageFacade<SyntaxKind> Language => VisualBasicFacade.Instance;
+
+    protected override bool IsExtensionMethod(SyntaxNode methodDeclaration) =>
+        methodDeclaration.Parent.Parent is ModuleBlockSyntax
+        && ((MethodStatementSyntax)methodDeclaration).AttributeLists
+            .SelectMany(x => x.Attributes)
+            .Any(IsExtension);
+
+    private bool IsExtension(AttributeSyntax attribute)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExtensionMethodShouldNotExtendObject.cs:41]: This is quite repetitive problem. On C# side, we have
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.CSharp/Extensions/AttributeSyntaxExtensions.cs

We could copy/paste it to VB project as well for symetry.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Thank you @Corniel for taking a look into this. I left some minor comments

**Comment by pavel-mikula-sonarsource** [comment]
I've pushed the RSPEC change so you can try to run the UTs

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        Sub ExtendsObject(obj As Object) ' Noncompliant {{Refactor this extension to extend a more concrete type.}}
+'           ^^^^^^^^^^^^^
+        End Sub
+        
+        <Extension>
+        Function ExtendsObject(obj As Object, other as Object) as Object ' Noncompliant
+            Return other
+        End Function
+		
+		<ExtensionAttribute>
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb:34]: No, that will not compile. I tried it. ;)

**Comment by Corniel** [comment]
Applied all.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb]: I just tried this and it compiled in empty solution:
```
    <System.Runtime.CompilerServices.ExtensionAttribute>
    Public Sub FullName(Arg As Object)
    End Sub

    <Runtime.CompilerServices.ExtensionAttribute>
    Public Sub ImplicitSystem(Arg As Object)
    End Sub

    <System.Runtime.CompilerServices.Extension>
    Public Sub ImplicitAttribute(Arg As Object)
    End Sub

    <Runtime.CompilerServices.Extension>
    Public Sub ImplicitSystemAndAttribute(Arg As Object)
    End Sub
```
Wasn't there some copy/paste issue?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+﻿Imports System.Runtime.CompilerServices
+
+Namespace S4225.ExtensionMethodShouldNotExtendObject
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb]: Historically, namespaces were part of the template that scaffolded a new rule. That's how it polluted the existing files. For most of the rules, it's useless noise that is not relevant for the purpose of the rule.

They make sense when the `namespace` or `using` matters. Or when the test case file is really big and we need to split it into topics and class names are not enough there.

In this rule, the test case was pretty weak and you've rewritten the whole thing. So it makes sense to clean the namespace too. I wouldn't do it on a bigger file.

File: analyzers/src/SonarAnalyzer.VisualBasic/Helpers/VisualBasicSyntaxHelper.cs [hunk 4]
```diff
@@ -187,7 +187,9 @@ private static bool IsOn(this ExpressionSyntax expression, SyntaxKind onKind)
             };
 
         public static string GetName(this SyntaxNode expression) =>
-            expression.GetIdentifier()?.ValueText ?? string.Empty;
+            expression is AttributeSyntax x
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.VisualBasic/Helpers/VisualBasicSyntaxHelper.cs:190]: This was not addressed

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExtensionMethodShouldNotExtendObject.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExtensionMethodShouldNotExtendObject.cs]: What I meant was to have `AttributeSyntaxExtensions.IsKnownType` on VB side, just as we have it on C#. And use that. The `IsExtension` is too specific to be an extension by itself.

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExtensionMethodShouldNotExtendObject.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+[DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+public sealed class ExtensionMethodShouldNotExtendObject : ExtensionMethodShouldNotExtendObjectBase<SyntaxKind, MethodStatementSyntax>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => VisualBasicFacade.Instance;
+
+    protected override bool IsExtensionMethod(MethodStatementSyntax declaration) =>
+        declaration.Parent.Parent is ModuleBlockSyntax
+        && declaration.AttributeLists
+            .SelectMany(x => x.Attributes)
+            .Any(x => x.IsExtension());
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExtensionMethodShouldNotExtendObject.cs:40]: This is now clean and easy to follow, I'd make it a single line.
```suggestion
        && declaration.AttributeLists.SelectMany(x => x.Attributes).Any(x => x.IsExtension());
```

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExtensionMethodShouldNotExtendObject.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+[DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+public sealed class ExtensionMethodShouldNotExtendObject : ExtensionMethodShouldNotExtendObjectBase<SyntaxKind, MethodStatementSyntax>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => VisualBasicFacade.Instance;
+
+    protected override bool IsExtensionMethod(MethodStatementSyntax declaration) =>
+        declaration.Parent.Parent is ModuleBlockSyntax
+        && declaration.AttributeLists
+            .SelectMany(x => x.Attributes)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/ExtensionMethodShouldNotExtendObject.cs:39]: Seing this, let's add test cases with multiple attributes and multi-attribute syntaxes:
```
    <Extension, DebuggerStepThrough()>
    Public Sub MoreAttributes1(Arg As Object)
    End Sub

    <Extension>
    <DebuggerStepThrough()>
    Public Sub MoreAttributes2(Arg As Object)
    End Sub
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb]: @pavel-mikula-sonarsource I agree. But I try to follow most of your conventions. I'm happy to drop noisy namespaces.

File: analyzers/src/SonarAnalyzer.VisualBasic/Helpers/VisualBasicSyntaxHelper.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.VisualBasic/Helpers/VisualBasicSyntaxHelper.cs]: I do not need it any more. Note that `AttributeSytnax` lacks an `Identifer` node.

**Comment by pavel-mikula-sonarsource** [comment]
I've resolved some conversations, there are few UTs to be added.

**Comment by pavel-mikula-sonarsource** [comment]
I've resolved some conversations, there are few UTs to be added.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb (hunk 2 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb]: I've also added these cases so we can move forward with this PR

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb [hunk 7]
```diff
@@ -46,4 +46,20 @@ Module Noncompliant
...
+    <System.Runtime.CompilerServices.ExtensionAttribute>
+    Public Sub FullName(Arg As Object)          ' Noncompliant
+    End Sub
+
+    <Runtime.CompilerServices.ExtensionAttribute>
+    Public Sub ImplicitSystem(Arg As Object)    ' Noncompliant
+    End Sub
+
+    <System.Runtime.CompilerServices.Extension>
+    Public Sub ImplicitAttribute(Arg As Object) ' Noncompliant
```
Comment (Corniel) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ExtensionMethodShouldNotExtendObject.vb:58]: Descriptive names! Good. That is not always the case in some older test case files.

**Comment by pavel-mikula-sonarsource** [comment]
@Corniel can you please rebase? We've (finally) deleted `RspecStrings.resx` file so we need to remove it from this PR too.

**Comment by Corniel** [comment]
Rebase done.

**Pull request merged**